soc/mediatek/mt8xxx/mtk_adsp_load.py

```diff
@@ -218,6 +218,93 @@ def le4(bstr):
     return struct.unpack("<I", bstr)[0]
 
 
+# Wrapper class for winstream logging.  Instantiate with a single
+# integer argument representing a local/in-process address for the
+# shared winstream memory.  The memory mapped access is encapsulated
+# with a Regs object for the fields and a ctypes array for the data
+# area.  The lockless algorithm in read() matches the C version in
+# upstream Zephyr, don't modify in isolation.  Note that on some
+# platforms word access to the data array (by e.g. copying a slice
+# into a bytes object or by calling memmove) produces bus errors
+# (plausibly an alignment requirement on the fabric with the DSP
+# memory, where arm64 python is happy doing unaligned loads?).  Access
+# to the data bytes is done bytewise for safety.
+class Winstream:
+    def __init__(self, addr):
+        r = Regs(addr)
+        r.WLEN = 0x00
+        r.START = 0x04
+        r.END = 0x08
+        r.SEQ = 0x0C
+        r.freeze()
+        # Sanity-check, the 32M size limit isn't a rule, but seems reasonable
+        if r.WLEN > 0x2000000 or (r.START >= r.WLEN) or (r.END >= r.WLEN):
+            raise RuntimeError("Invalid winstream")
+        self.regs = r
+        self.data = (ctypes.c_char * r.WLEN).from_address(addr + 16)
+        self.msg = bytearray(r.WLEN)
+        self.seq = 0
+
+    def read(self):
```

Comment:
This is yet another evolution of the python winstream code, based closely on @kv2019i 's most recent version with a few quality of life additions (like putting it into a stateful object that can track seq) and some optimization to the inner loop.   I didn't try to move it back into the cavstool code for lack of time, but it should probably work.

Unfortunately there's no easy way to share importable python code across "deployable" scripts without someone doing some build integration.  So we're stuck with cut/paste for now.

Reply:
The "just-scp-one-file" deployment model is very contagious :) But I think we need to make the jump sooner or later.

Reply:
FWIW it actually is possible to run python3 on a zip file that contains a whole tree of files.  And it's possible to construct a zip file with a "#!" shebang line in the first bytes to make it "executable" on the target (as long as there's a python interpreter).  And there's even a "zipapp" package in pip to automate that.  But we'd have to add some cmake integration to package it up at build time, and decide where to put the tree of modules, and extract them from the current users, etc... Some tedium and design decisions to be made, but it seems like a solvable problem.